juju/main.py

```diff
@@ -0,0 +1,156 @@
+            debugf(f'Removing entry {event_hook_path} as it does not point'
+                   ' to "install"')
+            event_hook_path.unlink()
+        else:
+            create_link = False
+
+    if create_link:
+        debugf(f'Creating a new relative symlink at {event_hook_path}'
+               f' to pointing to "install" located in {charm_dir}/hooks')
+        event_hook_path.symlink_to('install')
+
+
+def _setup_hooks(charm_dir, charm):
+    """Set up hooks for supported events.
+
+    Whether a charm can handle an event or not can be determined by
+    introspecting which events are bound to it.
+
+    Hooks are created as symlinks to "install" which may be either
+    a copy of main.py (which calls the main function) or a symlink
+    to it. Note that it is important to avoid creating a recursive
+    symlink which is why the install event itself is skipped.
+
+    charm_dir -- A root directory of the charm.
+    Charm -- A type of a concrete charm implementation.
+    CharmEvents -- A type declaring core Juju events.
+    Event -- A framework Event type.
+    BoundEvent -- A framework BoundEvent type.
+    """
+    for event_name in charm.on.events().keys():
+        if event_name != 'install':
+            _handle_event_link(charm_dir, event_name)
+
+
+def _emit_charm_event(charm, event_name):
+    """Emits a charm event based on a Juju event name.
+
+    charm -- A charm instance to emit an event from.
+    event_name -- A Juju event name to emit on a charm.
+    """
+    formatted_event_name = event_name.replace('-', '_')
+    event_to_emit = None
+    try:
+        event_to_emit = getattr(charm.on, formatted_event_name)
+    except AttributeError:
+        debugf(f"event {formatted_event_name} not defined for {charm}")
+
+    # If the event is not supported by the charm implementation, do
+    # not error out or try to emit it. This is to support rollbacks.
+    if event_to_emit is not None:
+        debugf(f'Emitting Juju event {event_name}')
+        event_to_emit.emit()
+
+
+def main():
+    """Setup the charm and dispatch the observed event.
+
+    The event name is based on the way this executable was called (argv[0]).
+    """
+    import charm as charm_module
+    import juju.charm
+    import juju.framework
+
+    charm_dir = _get_charm_dir()
+    juju_event_name = Path(sys.argv[0]).name
+
+    # TODO: If Juju unit agent crashes after exit(0) from the charm code
+    # the framework will commit the snapshot but Juju will not commit its
+    # operation.
+    charm_state_path = charm_dir / CHARM_STATE_FILE
+    framework = juju.framework.Framework(data_path=charm_state_path)
+    try:
+        # Set up charm-specific in-memory state.
+        charm = charm_module.Charm(framework, None, _load_metadata())
+
+        # Set up hooks on initial install or charm upgrade. Note that if
+        # a charm is force-upgraded this will only be called after
+        # upgrade-charm is executed.
```

Comment:
It only gets run if a unit is not in the error state, otherwise Juju just waits until you resolve a unit and then it runs the failed hook again followed by `config-changed`.

https://gist.github.com/dshcherb/c828894bd385ac4662e227d93c0ddcb7

I can make the comment more specific in saying that force-upgrade on a unit in the error state will not result in symlinks getting recreated.

The problem with checking for a revision is that I have to store state somewhere locally to compare against.

I could add config-changed to the list which will make it a more expensive operation as it will run more frequently. However, I don't think it's nearly as bad as executing it for relation events because config-changed is operator-initiated in most cases.

Reply:
Added config-changed to the new revision.

Reply:
I think storage hooks will need to be handled as well.

Judging by the code and public documentation it looks like "install" will run after any storage attachment events.

https://github.com/juju/juju/blob/692c32713224c80022aec33da220b3525efa582f/worker/uniter/storage/resolver.go#L78-L82

https://jaas.ai/docs/writing-charms-that-use-storage#heading--storage-hooks
> [name]-storage-attached hooks will be run before the install hook

Reply:
But the challenge with storage events is that hooks for them have to be there at the deployment time e.g.:

disks-storage-attached -> install

And the first event a unit agent will run will be `hooks/disks-storage-attached`, not `hooks/install`.

Reply:
I added a symlink for a storage endpoint called "disks" to the test charm and updated the `setup_hooks` test to run a storage event before the install event.

So, for now, the assumption is that a charm author will create an install symlink to `../lib/main.py` and any storage hook symlinks pointing to the `install` symlink.